### PR TITLE
[RISCV] Add Qualcomm uC Xqcicsr (CSR) extension

### PR DESCRIPTION
**Pull request by svs-quic**
The Qualcomm uC Xqcicsr extension adds 2 instructions that can read and write CSRs.

The current spec can be found at:
https://github.com/quic/riscv-unified-db/releases/latest

This patch adds assembler only support.

### CODE REVIEW

File: llvm/docs/RISCVUsage.rst [hunk 1]
```diff
@@ -329,6 +329,9 @@ The primary goal of experimental support is to assist in the process of ratifica
 ``experimental-smctr``, ``experimental-ssctr``
   LLVM implements the `1.0-rc3 specification <https://github.com/riscv/riscv-control-transfer-records/releases/tag/v1.0_rc3>`__.
 
+``experimental-xqcicsr``
```
Comment (topperc) [inline llvm/docs/RISCVUsage.rst:332]: This should be in the "Vendor Extensions" section further down

Comment (svs-quic) [reply llvm/docs/RISCVUsage.rst]: Thanks. I've moved it to the Vendor Extensions section.

**Review by lenary — APPROVED** [review]
I am out of the loop on the discussion in the call today, but I can confirm this is a correct implementation of Xqcicsr, according to the specification.

File: llvm/lib/TargetParser/RISCVISAInfo.cpp [hunk 2]
```diff
@@ -771,6 +771,10 @@ Error RISCVISAInfo::checkDependency() {
       return getIncompatibleError("xwchc", "zcb");
   }
 
+  if (Exts.count("xqcicsr") != 0 && (XLen != 32)) {
+    return getError("'xqcicsr' is only supported in 'rv32'");
```
Comment (topperc) [inline llvm/lib/TargetParser/RISCVISAInfo.cpp:775]: `in` -> `for` to match the similar error messages above.

File: llvm/test/MC/RISCV/xqcicsr-valid.s [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+# RUN:     | llvm-objdump --mattr=+experimental-xqcicsr --no-print-imm-hex -d - \
+# RUN:     | FileCheck -check-prefix=CHECK-INST %s
+
+# CHECK-INST: qc.csrrwr  a0, t0, s4
+# CHECK-ENC: encoding: [0x73,0x85,0x42,0x8d]
+qc.csrrwr  x10, x5, x20
+
+# CHECK-INST: qc.csrrwri s4, 31, a2
+# CHECK-ENC: encoding: [0x73,0x8a,0xcf,0x8e]
+qc.csrrwri x20, 31, x12
```
Comment (topperc) [inline llvm/test/MC/RISCV/xqcicsr-valid.s:19]: Missing new line

Comment (svs-quic) [reply llvm/test/MC/RISCV/xqcicsr-valid.s]: Done

File: llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+// Instructions
+//===----------------------------------------------------------------------===//
+
+let Predicates = [HasVendorXqcicsr, IsRV32], DecoderNamespace = "Xqcicsr" in {
+let hasSideEffects = 0, mayLoad = 0, mayStore = 0 in {
+  def QC_CSRRWR : RVInstR<0b1000110, 0b000, OPC_SYSTEM, (outs GPR:$rd),
+                          (ins GPR:$rs1, GPRNoX0:$rs2), "qc.csrrwr",
+                          "$rd, $rs1, $rs2">;
+
+  def QC_CSRRWRI : RVInstRBase<0b000, OPC_SYSTEM, (outs GPR:$rd),
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td:35]: I think you can just use `RVInstR` if you use `uimm5:$rs1`. It's what `CSR_ii` does for the standard CSR instructions.

Comment (svs-quic) [reply llvm/lib/Target/RISCV/RISCVInstrInfoXqci.td]: Done

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**